examples/tutorials/core_api_mnist/model_def_metrics.py

```diff
@@ -0,0 +1,210 @@
+# Report metrics to Determined
+
+from __future__ import print_function
+import argparse
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+import torch.optim as optim
+from torchvision import datasets, transforms
+from torch.optim.lr_scheduler import StepLR
+
+# NEW: import Determined
+import determined as det
+
+
+class Net(nn.Module):
+    def __init__(self):
+        super(Net, self).__init__()
+        self.conv1 = nn.Conv2d(1, 32, 3, 1)
+        self.conv2 = nn.Conv2d(32, 64, 3, 1)
+        self.dropout1 = nn.Dropout(0.25)
+        self.dropout2 = nn.Dropout(0.5)
+        self.fc1 = nn.Linear(9216, 128)
+        self.fc2 = nn.Linear(128, 10)
+
+    def forward(self, x):
+        x = self.conv1(x)
+        x = F.relu(x)
+        x = self.conv2(x)
+        x = F.relu(x)
+        x = F.max_pool2d(x, 2)
+        x = self.dropout1(x)
+        x = torch.flatten(x, 1)
+        x = self.fc1(x)
+        x = F.relu(x)
+        x = self.dropout2(x)
+        x = self.fc2(x)
+        output = F.log_softmax(x, dim=1)
+        return output
+
+
+# NEW: modified function header to include core_context for metric reporting
+def train(args, model, device, train_loader, optimizer, epoch, core_context):
+    model.train()
+    for batch_idx, (data, target) in enumerate(train_loader):
+        data, target = data.to(device), target.to(device)
+        optimizer.zero_grad()
+        output = model(data)
+        loss = F.nll_loss(output, target)
+        loss.backward()
+        optimizer.step()
+        if (batch_idx + 1) % args.log_interval == 0:
+
+            print(
+                "Train Epoch: {} [{}/{} ({:.0f}%)]\tLoss: {:.6f}".format(
+                    epoch,
+                    (batch_idx) * len(data),
+                    len(train_loader.dataset),
+                    100.0 * (batch_idx) / len(train_loader),
+                    loss.item(),
+                )
+            )
+
+            # NEW: report training metrics to Determined master via core context.
+            # Index by (batch_idx + 1) * (epoch-1) * len(train_loader) to continuously plot loss on one graph for consecutive epochs.
+            core_context.train.report_training_metrics(
+                steps_completed=(batch_idx + 1) + (epoch - 1) * len(train_loader),
+                metrics={"train_loss": loss.item()},
+            )
+
+            if args.dry_run:
+                break
+
+
+# NEW: modified function header to include args, epoch, test_loader, core_context for metric reporting and a steps_completed parameter to plot metrics
+def test(args, model, device, test_loader, epoch, core_context, steps_completed):
+    model.eval()
+    test_loss = 0
+    correct = 0
+    with torch.no_grad():
+        for batch_idx, (data, target) in enumerate(test_loader):
```

Comment:
you changed this line, but I don't see that `batch_idx` is actually used anywhere